lib/Conversion/ArithExtToArith/ArithExtToArith.td

```diff
@@ -33,4 +33,121 @@ def ConvertSubIfGE : Pattern<
   ]
 >;
 
+
+def HasEnoughSpaceAddSub: Constraint<CPred<"llvm::cast<IntegerType>(getElementTypeOrSelf($_self.getType())).getWidth() >= ($0.getValue() - 1).getActiveBits() + 1">,
+"underlying type is sufficient for modular add/sub operation without overflow">;
+
+def HasEnoughSpaceMul: Constraint<CPred<"llvm::cast<IntegerType>(getElementTypeOrSelf($_self.getType())).getWidth() >= 2 * ($0.getValue() - 1).getActiveBits()">,
+ "underlying type is sufficient for modular mul operation without overflow">;
+
+def CastModulusAttributeAddSub : NativeCodeCall<"modulusHelper($0,$1,false)">;
+def CastModulusAttributeMul : NativeCodeCall<"modulusHelper($0,$1,true)">;
+
+def ConvertAddSimple : Pattern<
+  (ArithExt_AddOp:$op $x, $y, $mod),
+  [
+    (Arith_AddIOp:$add $x, $y, DefOverflow),
+    (Arith_RemUIOp $add, (Arith_ConstantOp (CastModulusAttributeAddSub $mod, $x)))
+  ],
+  [(HasEnoughSpaceAddSub:$op $mod)],
+  [],
+  (addBenefit 2)
+>;
+
+def ConvertSubSimple : Pattern<
+  (ArithExt_SubOp:$op $x, $y, $mod),
+  [
+    (Arith_SubIOp:$sub $x, $y, DefOverflow),
+    (Arith_RemUIOp $sub, (Arith_ConstantOp (CastModulusAttributeAddSub $mod, $x)))
+  ],
+  [(HasEnoughSpaceAddSub:$op $mod)],
+  [],
+  (addBenefit 2)
+>;
+
+def ConvertMulSimple : Pattern<
+  (ArithExt_MulOp:$op $x, $y, $mod),
+  [
+    (Arith_MulIOp:$mul $x, $y, DefOverflow),
+    (Arith_RemUIOp $mul, (Arith_ConstantOp (CastModulusAttributeMul $mod, $x)))
+  ],
+  [(HasEnoughSpaceMul:$op $mod)],
+  [],
+  (addBenefit 2)
+>;
+
+def ConvertMacSimple : Pattern<
+  (ArithExt_MacOp:$op $x, $y, $acc, $mod),
+  [
+    (Arith_MulIOp:$mul $x, $y, DefOverflow),
+    (Arith_AddIOp:$add $mul, $acc, DefOverflow),
+    (Arith_RemUIOp $add, (Arith_ConstantOp (CastModulusAttributeMul $mod, $x)))
+  ],
+  [(HasEnoughSpaceMul:$op $mod)],
+  [],
+  (addBenefit 2)
+>;
+
+def ConvertAdd : Pattern<
+  (ArithExt_AddOp $x, $y, $mod),
+  [
+    (Arith_ConstantOp:$newmod (CastModulusAttributeAddSub $mod, $x)),
+    (Arith_AddIOp:$add
+      (Arith_ExtUIOp $x,
+        (returnType $newmod)),
+      (Arith_ExtUIOp $y,
+        (returnType $newmod)),
+      DefOverflow),
+    (Arith_TruncIOp:$res
+      (Arith_RemUIOp $add, $newmod))
+  ]
+>;
+
+def ConvertSub : Pattern<
+  (ArithExt_SubOp $x, $y, $mod),
+  [
+    (Arith_ConstantOp:$newmod (CastModulusAttributeAddSub $mod, $x)),
+    (Arith_SubIOp:$sub
+      (Arith_ExtUIOp $x,
+        (returnType $newmod)),
+      (Arith_ExtUIOp $y,
+        (returnType $newmod)),
+      DefOverflow),
+    (Arith_TruncIOp:$res
+      (Arith_RemUIOp $sub, $newmod))
+  ]
+>;
+
+def ConvertMul : Pattern<
+  (ArithExt_MulOp $x, $y, $mod),
+  [
+    (Arith_ConstantOp:$newmod (CastModulusAttributeMul $mod, $x)),
+    (Arith_MulIOp:$mul
+      (Arith_ExtUIOp $x,
+        (returnType $newmod)),
+      (Arith_ExtUIOp $y,
+        (returnType $newmod)),
+      DefOverflow),
+    (Arith_TruncIOp:$res
+      (Arith_RemUIOp $mul, $newmod))
+  ]
+>;
+
+def ConvertMac : Pattern<
```

Comment:
Just curious, to me it seems like you could express `arith_ext.mac` as an `arith_ext.mul` and then `arith_ext.add`, and then utilize the patterns above for lowering to arith.

But like the pattern below, we can save on a `remui`-`trunci` sequence on the intermediate mul with the pattern below. Was that the intention to create a mac operation?

Reply:
The main reason is actually that `mac` is a native HW instruction in (some? all?) RLWE accelerators, but I also like having it for the efficiency gain ;)  

Though I guess we could also achieve saving the `remui`/etc just with patterns that apply to `add(mul, _)`, but that might make things a bit messy with all the "checking if bits fit" logic.

Reply:
Yeah, it seems pretty messy to try to "lazy remui".
Thank you!